pkg/operation/botanist/etcdencryption.go

```diff
@@ -0,0 +1,321 @@
+func (b *Botanist) writeEncryptionConfigurationSecretToGarden(ecYamlBytes []byte) error {
+	secretNameInGardenCluster := b.calculateEtcdEncryptionSecretNameInGardenCluster()
+	secretObj := b.createEncryptionConfigurationSecret(secretNameInGardenCluster, b.Shoot.Info.Namespace, ecYamlBytes, false)
+	secretObj.ObjectMeta.OwnerReferences = []metav1.OwnerReference{
+		*metav1.NewControllerRef(b.Shoot.Info, gardenv1beta1.SchemeGroupVersion.WithKind("Shoot")),
+	}
+	if _, err := b.K8sGardenClient.CreateSecretObject(secretObj, true); err != nil {
+		return err
+	}
+	return nil
+}
+
+// isEncryptionConfigurationConsistent checks whether the configuration is consistent in seed and garden
+func (b *Botanist) isEncryptionConfigurationConsistent(ecSeed *apiserverconfigv1.EncryptionConfiguration, ecGarden *apiserverconfigv1.EncryptionConfiguration) (bool, error) {
+	equal, err := encryptionconfiguration.Equals(ecSeed, ecGarden)
+	if err != nil {
+		return false, fmt.Errorf("error when comparing EncryptionConfiguration in seed cluster and garden cluster: %v", err)
+	}
+	if !equal {
+		return false, nil
+	}
+	consistent, err := encryptionconfiguration.IsConsistent(ecSeed)
+	if err != nil {
+		return false, fmt.Errorf("EncryptionConfiguration in seed cluster is not consistent: %v", err)
+	}
+	if !consistent {
+		return false, nil
+	}
+	return true, nil
+}
+
+// needToRewriteShootSecrets checks whether the secrets in the shoot need to
+// be rewritten, e.g. after a change to the EncryptionConfiguration
+func (b *Botanist) needToRewriteShootSecrets() (needToRewriteSecrets bool, err error) {
+	ecs, err := b.K8sSeedClient.GetSecret(b.Operation.Shoot.SeedNamespace, common.EtcdEncryptionSecretName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return false, fmt.Errorf("no EncryptionConfiguration found in seed cluster (%v)", b.Operation.Shoot.SeedNamespace)
+		}
+		return false, err
+	}
+	if ecs.Annotations == nil {
+		// no annotations found, rewrite secrets
+		return true, nil
+	}
+	lastSecretRewriteChecksum, ok := ecs.Annotations[common.EtcdEncryptionChecksumAnnotationName]
+	if !ok {
+		// no annotation found, rewrite secrets
+		return true, nil
+	}
+	if b.CheckSums[common.EtcdEncryptionSecretName] == lastSecretRewriteChecksum {
+		// no change in checksum, no need to rewrite secrets
+		return false, nil
+	}
+	// checksum changed, do rewrite
+	return true, nil
+}
+
+// rewriteShootSecrets rewrites all secrets of the shoot.
+// This will take into account the current EncryptionConfiguration.
+func (b *Botanist) rewriteShootSecrets() error {
+	secretList, err := b.K8sShootClient.ListSecrets(metav1.NamespaceAll, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, secret := range secretList.Items {
+		secretName := secret.GetName()
+		secretNamespace := secret.GetNamespace()
+		attempt := 0
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			attempt++
+			_, updateErr := b.K8sShootClient.UpdateSecretObject(&secret)
```

Comment:
Don't use deprecated methods - use the controller-runtime client instead. 
Other question: Do you need to do a full-fledged `Update` call or would a 0-length patch also suffice?